pkg/check/table_structure.go

```diff
@@ -77,73 +156,283 @@ func (c *TablesChecker) Name() string {
+func (c *ShardingTablesCheck) Check(ctx context.Context) *Result {
+	r := &Result{
+		Name:  c.Name(),
+		Desc:  "check consistency of sharding table structures",
+		State: StateSuccess,
+		Extra: fmt.Sprintf("sharding %s", c.name),
+	}
+
+	var (
+		hasAutoIncrementKeyTableNum int
+		hasAutoIncrementKeyTables   = make(map[string][]string)
+
+		stmtNode  *ast.CreateTableStmt
+		tableName string
+	)
+	for instance, schemas := range c.tables {
+		db, ok := c.dbs[instance]
+		if !ok {
+			markCheckError(r, errors.NotFoundf("client for instance %s", instance))
+			return r
+		}
+
+		for schema, tables := range schemas {
+			for _, table := range tables {
+				if len(tables) == 0 {
+					continue
+				}
+				statement, err := dbutil.GetCreateTableSQL(ctx, db, schema, table)
+				if err != nil {
+					markCheckError(r, err)
+					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
+					return r
+				}
+
+				stmt, err := parser.New().ParseOneStmt(statement, "", "")
+				if err != nil {
+					markCheckError(r, errors.Annotatef(err, "statement %s", statement))
+					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
+					return r
+				}
+
+				ctStmt, ok := stmt.(*ast.CreateTableStmt)
+				if !ok {
+					markCheckError(r, errors.Errorf("Expect CreateTableStmt but got %T", stmt))
+					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
+					return r
+				}
+
+				if c.hashAutoIncrementKey(ctStmt) {
+					hasAutoIncrementKeyTableNum++
+					hasAutoIncrementKeyTables[instance] = append(hasAutoIncrementKeyTables[instance], dbutil.TableName(schema, table))
+				}
+
+				if stmtNode == nil {
+					stmtNode = ctStmt
+					tableName = dbutil.TableName(schema, table)
+					continue
+				}
+
+				err = c.checkConsistency(stmtNode, ctStmt, tableName, dbutil.TableName(schema, table))
+				if err != nil {
+					markCheckError(r, err)
+					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
+					r.Instruction = "please set same table structure for sharding tables"
+					return r
+				}
+			}
+		}
+	}
+
+	if hasAutoIncrementKeyTableNum > 1 {
+		r.State = StateWarning
+		r.ErrorMsg = fmt.Sprintf("(%d) tables %+v have auto-increment key, would conflict with each other to cause data corruption", hasAutoIncrementKeyTableNum, hasAutoIncrementKeyTables)
+		r.Instruction = "please set column mapping rules for them, or handle it by yourself"
+		r.Extra = AutoIncrementKeyChecking
+	}
+
+	return r
+}
+
+func (c *ShardingTablesCheck) hashAutoIncrementKey(stmt *ast.CreateTableStmt) bool {
+	for _, col := range stmt.Cols {
+		var (
+			hasAutoIncrementOpt bool
+			isUnique            bool
+		)
+		for _, opt := range col.Options {
+			switch opt.Tp {
+			case ast.ColumnOptionAutoIncrement:
+				hasAutoIncrementOpt = true
+			case ast.ColumnOptionPrimaryKey, ast.ColumnOptionUniqKey:
+				isUnique = true
+			}
+		}
+
+		if hasAutoIncrementOpt || isUnique {
+			return true
```

Comment:
can `return true` just in L352 and L354 ?

Reply:
it's auto increment key, should be `hasAutoIncrementOpt && isUnique`